pkg/controller/issueapi/validate_code.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/google/exposure-notifications-verification-server/pkg/sms"
+)
+
+// populateCode populates and validates a code from an issue request.
+func (c *Controller) populateCode(ctx context.Context, request *api.IssueCodeRequest,
+	authApp *database.AuthorizedApp, membership *database.Membership, realm *database.Realm) (*database.VerificationCode, *issueResult) {
+	logger := logging.FromContext(ctx).Named("issueapi.populateCode")
+
+	now := time.Now().UTC()
+	vCode := &database.VerificationCode{
+		RealmID:           realm.ID,
+		IssuingExternalID: request.ExternalIssuerID,
+		TestType:          strings.ToLower(request.TestType),
+		UUID:              project.TrimSpaceAndNonPrintable(request.UUID),
+		ExpiresAt:         now.Add(realm.CodeDuration.Duration),
+		LongExpiresAt:     now.Add(realm.LongCodeDuration.Duration),
+	}
+	if membership != nil {
+		vCode.IssuingUserID = membership.UserID
+	}
+	if authApp != nil {
+		vCode.IssuingAppID = authApp.ID
+	}
+
+	// If this realm requires a date but no date was specified, return an error.
+	if realm.RequireDate && request.SymptomDate == "" && request.TestDate == "" {
+		return nil, &issueResult{
+			obsResult:   observability.ResultError("MISSING_REQUIRED_FIELDS"),
+			httpCode:    http.StatusBadRequest,
+			errorReturn: api.Errorf("missing either test or symptom date").WithCode(api.ErrMissingDate),
+		}
+	}
+
+	// Parse SymptomDate and TestDate
+	var result *issueResult
+	vCode.SymptomDate, result = c.parseDate(request.SymptomDate, int(request.TZOffset), &onsetSettings)
+	if result != nil {
+		return nil, result
+	}
+	vCode.TestDate, result = c.parseDate(request.TestDate, int(request.TZOffset), &testSettings)
+	if result != nil {
+		return nil, result
+	}
+
+	// Verify SMS configuration if phone was provided
+	var smsProvider sms.Provider
+	if request.Phone != "" {
+		smsProvider, err := realm.SMSProvider(c.db)
+		if err != nil {
+			logger.Errorw("failed to get sms provider", "error", err)
+			return nil, &issueResult{
+				obsResult:   observability.ResultError("FAILED_TO_GET_SMS_PROVIDER"),
+				httpCode:    http.StatusInternalServerError,
+				errorReturn: api.Errorf("failed to get sms provider").WithCode(api.ErrInternal),
+			}
+		}
+		if smsProvider == nil {
+			err := fmt.Errorf("phone provided, but no sms provider is configured")
+			return nil, &issueResult{
+				obsResult:   observability.ResultError("FAILED_TO_GET_SMS_PROVIDER"),
+				httpCode:    http.StatusBadRequest,
+				errorReturn: api.Error(err),
+			}
+		}
+	}
+
+	if request.Phone == "" || smsProvider == nil {
+		// If this isn't going to be send via SMS, make the long code expiration time same as short.
+		// This is because the long code will never be shown or sent.
+		vCode.LongExpiresAt = vCode.ExpiresAt
+	}
+
+	// If there is a client-provided UUID, check if a code has already been issued.
+	// this prevents us from consuming quota on conflict.
+	if vCode.UUID != "" {
```

Comment:
Style nit: I'd prefer if we _didn't_ set UUID on the vcode above and instead set it here. This is a little unclear because you're calling `vCode.UUID`, but you're actually asserting the `request.UUID`. I'd prefer:

```go
if uuid := project.TrimSpaceNotPrintable(request.UUID); uuid != "" {
  if _, err := realm.FindVerificationCodeByUUID(c.db, uuid); err != nil {
    if database.IsNotFound(err) {
      vcode.UUID = uuid
    }
    return err
  }

  // no error, meaning record with UUID exists, conflict
  return "UUID_CONFLICT"
}
```